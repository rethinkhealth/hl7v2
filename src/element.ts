import { SEQUENCE_STARTING_INDEX } from "./constants";
import { IDelimiters } from "./delimiters";

export interface IElement {
  sequence: string;
  toJson(): any;
  value: string | IElement[];
}

export interface ElementOptions {
  delimiters?: IDelimiters;
}

export class Element implements IElement {
  private readonly options: ElementOptions | undefined;
  private readonly raw: string;

  public readonly sequence: string;

  private _value: string | IElement[];
  public get value(): string | IElement[] {
    return this._value;
  }

  public get isRepeatField(): boolean {
    if (!this.options?.delimiters) return false;
    else return this.raw.includes(this.options.delimiters.repeatSeparator);
  }

  constructor(value: string, sequence: string, options?: ElementOptions) {
    this.raw = value;
    this.sequence = sequence;
    this.options = options;

    // Setup the initial value
    this._value = this.raw;

    if (this.options?.delimiters) {
      if (this.isRepeatField) {
        this.parseValueForSeparator(this.options.delimiters.repeatSeparator);
      } else {
        this.parseValueForSeparator(this.options.delimiters.componentSeparator);
      }
    }
  }

  public toJson() {
    if (typeof this.value === "string") {
      return this.value;
    } else {
      // For repeating fields, we need to return an array of objects
      if (this.isRepeatField) {
        return this.value.map((element) => element.toJson());
      } else {
        const response = {} as any;
        this.value.forEach((element) => {
          response[element.sequence] = element.toJson();
        });
        return response;
      }
    }
  }

  // !Private setup
  private parseValueForSeparator(separator: string): boolean {
    if (this.raw.includes(separator)) {
      const values = this.raw.split(separator);
      this._value = values.map(
        (value, index) =>
          new Element(
            value,
            (index + SEQUENCE_STARTING_INDEX).toString(),
            this.options
          )
      );
      return true;
    }
    return false;
  }
}
