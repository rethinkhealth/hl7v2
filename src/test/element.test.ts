import { Element, IElement } from "../element";

describe("Element", () => {
  it("should return the default value if simple string", () => {
    // Given
    const value = "this is a value";
    const sequence = "PID.1";

    // When
    const element = new Element(value, sequence);

    // Then
    expect(element.value).toBe(value);
    expect(element.sequence).toBe(sequence);
  });

  it("should return components value if component separator exists", () => {
    // Given
    const value = "Bond^James^^007";
    const sequence = "PID.5";

    // When
    const element = new Element(value, sequence);

    // Then
    expect(element.sequence).toEqual("PID.5");
    expect(element.value.length).toEqual(4);
    expect((element.value as IElement[])[0].value).toEqual("Bond");
    expect((element.value as IElement[])[0].sequence).toEqual("PID.5.1");
  });

  it("should return repeated values if repeated separator", () => {
    const value =
      "(919)007-0007^^^^^919^0070007~(777)707-0707^^CP^^^777^7070707~^NET^X.400^007@BritishSecretService.com";
    const sequence = "PID.5.12";

    // When
    const element = new Element(value, sequence);

    // Then
    expect(element.sequence).toEqual("PID.5.12");
    expect(element.value.length).toEqual(3);
    expect(element.value).toMatchSnapshot();
  });
});

describe("Element JSON formatting", () => {
  it("should return simple value", () => {
    const value = "simple.value.com";
    const sequence = "PID.5.10";

    // When
    const element = new Element(value, sequence);

    // Then
    expect(element.toJson()).toEqual(value);
  });

  it("should return nested object for composed element", () => {
    const value =
      "(919)007-0007^^^^^919^0070007~(777)707-0707^^CP^^^777^7070707~^NET^X.400^007@BritishSecretService.com";
    const sequence = "PID.5.10";

    // When
    const element = new Element(value, sequence);

    // Then
    expect(element.toJson()).toEqual({
      "PID.5.10.1": {
        "PID.5.10.1.1": "(919)007-0007",
        "PID.5.10.1.2": "",
        "PID.5.10.1.3": "",
        "PID.5.10.1.4": "",
        "PID.5.10.1.5": "",
        "PID.5.10.1.6": "919",
        "PID.5.10.1.7": "0070007",
      },
      "PID.5.10.2": {
        "PID.5.10.2.1": "(777)707-0707",
        "PID.5.10.2.2": "",
        "PID.5.10.2.3": "CP",
        "PID.5.10.2.4": "",
        "PID.5.10.2.5": "",
        "PID.5.10.2.6": "777",
        "PID.5.10.2.7": "7070707",
      },
      "PID.5.10.3": {
        "PID.5.10.3.1": "",
        "PID.5.10.3.2": "NET",
        "PID.5.10.3.3": "X.400",
        "PID.5.10.3.4": "007@BritishSecretService.com",
      },
    });
  });
});
