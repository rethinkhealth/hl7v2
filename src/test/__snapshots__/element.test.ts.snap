// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element should return repeated values if repeated separator 1`] = `
[
  Element {
    "_value": [
      Element {
        "_value": "(919)007-0007",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "(919)007-0007",
        "sequence": "PID.5.12.1.1",
      },
      Element {
        "_value": "",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "",
        "sequence": "PID.5.12.1.2",
      },
      Element {
        "_value": "",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "",
        "sequence": "PID.5.12.1.3",
      },
      Element {
        "_value": "",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "",
        "sequence": "PID.5.12.1.4",
      },
      Element {
        "_value": "",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "",
        "sequence": "PID.5.12.1.5",
      },
      Element {
        "_value": "919",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "919",
        "sequence": "PID.5.12.1.6",
      },
      Element {
        "_value": "0070007",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "0070007",
        "sequence": "PID.5.12.1.7",
      },
    ],
    "options": {
      "delimiters": {
        "componentSeparator": "^",
        "escapeSeparator": "\\",
        "fieldSeparator": "|",
        "repeatSeparator": "~",
        "subComponentSeparator": "&",
        "terminator": "
",
      },
    },
    "raw": "(919)007-0007^^^^^919^0070007",
    "sequence": "PID.5.12.1",
  },
  Element {
    "_value": [
      Element {
        "_value": "(777)707-0707",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "(777)707-0707",
        "sequence": "PID.5.12.2.1",
      },
      Element {
        "_value": "",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "",
        "sequence": "PID.5.12.2.2",
      },
      Element {
        "_value": "CP",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "CP",
        "sequence": "PID.5.12.2.3",
      },
      Element {
        "_value": "",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "",
        "sequence": "PID.5.12.2.4",
      },
      Element {
        "_value": "",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "",
        "sequence": "PID.5.12.2.5",
      },
      Element {
        "_value": "777",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "777",
        "sequence": "PID.5.12.2.6",
      },
      Element {
        "_value": "7070707",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "7070707",
        "sequence": "PID.5.12.2.7",
      },
    ],
    "options": {
      "delimiters": {
        "componentSeparator": "^",
        "escapeSeparator": "\\",
        "fieldSeparator": "|",
        "repeatSeparator": "~",
        "subComponentSeparator": "&",
        "terminator": "
",
      },
    },
    "raw": "(777)707-0707^^CP^^^777^7070707",
    "sequence": "PID.5.12.2",
  },
  Element {
    "_value": [
      Element {
        "_value": "",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "",
        "sequence": "PID.5.12.3.1",
      },
      Element {
        "_value": "NET",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "NET",
        "sequence": "PID.5.12.3.2",
      },
      Element {
        "_value": "X.400",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "X.400",
        "sequence": "PID.5.12.3.3",
      },
      Element {
        "_value": "007@BritishSecretService.com",
        "options": {
          "delimiters": {
            "componentSeparator": "^",
            "escapeSeparator": "\\",
            "fieldSeparator": "|",
            "repeatSeparator": "~",
            "subComponentSeparator": "&",
            "terminator": "
",
          },
        },
        "raw": "007@BritishSecretService.com",
        "sequence": "PID.5.12.3.4",
      },
    ],
    "options": {
      "delimiters": {
        "componentSeparator": "^",
        "escapeSeparator": "\\",
        "fieldSeparator": "|",
        "repeatSeparator": "~",
        "subComponentSeparator": "&",
        "terminator": "
",
      },
    },
    "raw": "^NET^X.400^007@BritishSecretService.com",
    "sequence": "PID.5.12.3",
  },
]
`;
